packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,172 @@
+  defaultProps?: Partial<Props>;
+};
+
+/**
+ * Creates a slot from a slot shorthand or properties (`props.SLOT_NAME` or `props` itself)
+ *
+ * @param options - optional values you can pass to alter the signature of a slot, those values are:
+ *
+ * * `elementType` - the base element type of a slot, defaults to `'div'`
+ * * `defaultProps` - similar to a React component declaration, you can provide a slot default properties to be merged with the shorthand/properties provided
+ * * `required` - a boolean that indicates if a slot will be rendered even if it's base value is `undefined`.
+ * By default if `props.SLOT_NAME` is `undefined` then `state.SLOT_NAME` becomes `undefined`
+ * and nothing will be rendered, but if `required = true` then `state.SLOT_NAME` becomes an object
+ * with the values provided by `options.defaultProps` (or `{}`). This is useful for cases such as providing a default content
+ * in case no shorthand is provided, like the case of the `expandIcon` slot for the `AccordionHeader`
+ *
+ * @example of a required nullable slot
+ * ```tsx
+ * // AccordionHeader.types.ts
+ * type AccordionHeaderSlots = {
+ *    expandIcon?: Slot<'span'>
+ * }
+ * // useAccordionHeader.ts
+ * const state = {
+ *  expandIcon: slot(expandIconShorthand, {
+ *    required: true,
+ *    elementType: 'span',
+ *    defaultProps: { children: <ChevronRight/>, 'aria-hidden': true}
+ *  })
+ * }
+ * // renderAccordionHeader
+ * state.expandIcon && <state.expandIcon/>
+ * ```
+ */
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined,
+  options: { required: true } & SlotOptions<Props>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps>(
+  value: SlotComponent<Props>,
+  options?: { required: true } & Partial<SlotOptions<Props>>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & Partial<SlotOptions<Props>> = {},
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, elementType } = options;
+
+  if (value === null || (value === undefined && !required)) {
+    return undefined;
+  }
+
+  let metadata: SlotComponentMetadata<Props>;
+  if (isSlot<Props>(value)) {
+    metadata = value[SLOT_COMPONENT_METADATA_SYMBOL];
+    if (elementType !== undefined) {
+      metadata.elementType = elementType;
+    }
+  } else if (elementType !== undefined) {
+    metadata = { elementType };
+  } else {
+    throw new Error("[react-utilities]: slot options.elementType is required when value isn't a slot itself");
```

Comment:
nit: let's wrap this with env check to not throw in production

Reply:
This is a bit of a problem. As it was discussed before it would be nice to explicitly require `elementType`,  this is what this throw clause is doing, to remove it from production we would have to provide a default `elementType` value?!

Reply:
My proposal is to fallback to "div" in production and throw in dev